rfcs/2023-radix-form-primitive.md

```diff
@@ -0,0 +1,301 @@
+        event.preventDefault();
+        const data = Object.fromEntries(new FormData(event.currentTarget));
+
+        // maybe do an async server call and gather some errors from the server to display
+        submitForm(data)
+        .then(() => …)
+        /**
+         * Map errors from your server response into the expected format of `serverErrors` (see type below)
+         * For example:
+         * {
+         *   email: [ { code: 'email_invalid', 'This is not a valid email.' } ],
+         *   global: [ { code: 'server_error', 'Something went wrong.' }]
+         * }
+         */
+        .catch((errors) => setServerErrors(mapServerErrors(errors)));
+      }}
+      serverErrors={serverErrors}
+      onServerErrorsChange={setServerErrors}
+    >
+      <Form.Field name="name">
+        <Form.Label>Full name</Form.Label>
+        <Form.Control />
+        <Form.ClientMessage type="missingValue">Please enter your name.</Form.ClientMessage>
+        <Form.ServerMessage />
+      </Form.Field>
+
+      <Form.Field name="email">
+        <Form.Label>Email address</Form.Label>
+        <Form.Control type="email" />
+        <Form.ClientMessage type="missingValue">Please enter your email.</Form.ClientMessage>
+        <Form.ClientMessage type="typeMismatch">Please provide a valid email.</Form.ClientMessage>
+        <Form.ServerMessage /> {/* will yield "This is not a valid email." */}
+      </Form.Field>
+
+      <Form.Field name="age">
+        <Form.Label>Age</Form.Label>
+        <Form.Control type="number" min="18" max="57" step="1" />
+        <Form.ClientMessage type="missingValue">Please enter your age.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeOverflow">You must be 18 or older.</Form.ClientMessage>
+        <Form.ClientMessage type="rangeUnderflow">You must be 57 or younger.</Form.ClientMessage>
+        <Form.ClientMessage type="stepMismatch">Please enter a whole number.</Form.ClientMessage>
+        <Form.ServerMessage />
+      </Form.Field>
+
+      <Form.Submit>Submit</Form.Submit>
+
+      <Form.ServerMessage /> {/* will yield "Something went wrong." */}
+    </Form.Root>
+  );
+}
+```
+
+If no `children` are provided to `Form.ServerError`, it will display a concatenation of error messages from the server. If passing a React node, it will render this node instead. For finer control, you can pass a render function instead which will be called with the server errors.
+
+The server messages will be routed to the correct fields, based on the `name` attribute on `Form.Field` matching in the `serverErrors` object (see type below). There is also a `global` key which will be used for global errors (displayed in the `Form.ServerMessage` sitting outside a `Form.Field`).
+
+```ts
+interface ServerError {
+  code: string;
+  message: React.ReactNode;
+}
+
+type ServerErrors = {
+  [fieldName in string]?: ServerError[];
+} & {
+  global?: ServerError[];
+};
```

Comment:
These types are sometimes annoying, the other option here is:

```ts
type ServerErrors = {
  form: ServerError[],
  fields: Record<string, ServerError[]>,
}

// You could make it optional in the prop so that it's easier to pass in
type ServerErrorsProp = Partial<ServerErrors>
```

Reply:
Interesting. What specifically is annoying with this type?